File: js/src/analytics/walkforward.js
/**
 * walkforward.js — anchored + rolling walk-forward harnesses.
 * backtest_fn may return a plain object or a Promise.
 * Spec §4 (Walk-forward section).
 */

/**
 * Anchored walk-forward.
 * Train slice always starts at bar 0; test slice follows immediately.
 *
 * @param {function} backtestFn  (trainBars, testBars) -> {train, test} or Promise
 * @param {any[]}    bars        full bar array
 * @param {number}   nFolds      number of folds
 * @param {number}   trainMinBars minimum bars for first training window
 * @returns {Promise<Array<{fold:number, trainBars:any[], testBars:any[], result:any}>>}
 */
export async function walkforwardAnchored(backtestFn, bars, nFolds, trainMinBars) {
    const n = bars.length;
    if (nFolds <= 0 || n === 0) return [];
    const results = [];
    const extra = n - trainMinBars;

    for (let k = 0; k < nFolds; k++) {
        const trainEnd = trainMinBars + Math.floor(k * extra / nFolds);
        const testLen  = Math.floor(extra / nFolds);
        const testStart = trainEnd;
        const testEnd   = testStart + testLen;

        if (testEnd > n) break;
        if (trainEnd <= 0 || testLen <= 0) continue;

        const trainBars = bars.slice(0, trainEnd);
        const testBars  = bars.slice(testStart, testEnd);

        const result = await Promise.resolve(backtestFn(trainBars, testBars));
        results.push({ fold: k, trainBars, testBars, result });
    }
    return results;
}

/**
 * Rolling walk-forward.
 * Sliding train/test windows stepped by `step` bars.
 *
 * @param {function} backtestFn  (trainBars, testBars) -> {train, test} or Promise
 * @param {any[]}    bars
 * @param {number}   trainWindow  number of bars in training window
 * @param {number}   testWindow   number of bars in test window
 * @param {number}   step         bars to advance each fold
 * @returns {Promise<Array<{fold:number, trainStart:number, testStart:number, result:any}>>}
 */
export async function walkforwardRolling(backtestFn, bars, trainWindow, testWindow, step) {
    const n = bars.length;
    if (n === 0 || step <= 0) return [];
    const results = [];
    let fold = 0;

    for (let i = 0; i + trainWindow + testWindow <= n; i += step) {
        const trainBars = bars.slice(i, i + trainWindow);
        const testBars  = bars.slice(i + trainWindow, i + trainWindow + testWindow);
        const result = await Promise.resolve(backtestFn(trainBars, testBars));
        results.push({
            fold,
            trainStart: i,
            testStart:  i + trainWindow,
            result,
        });
        fold++;
    }
    return results;
}

File: cpp/src/analytics/walkforward.cpp
/**
 * AlgoForge — src/analytics/walkforward.cpp
 * Anchored and rolling walk-forward harnesses.
 */

#include "core/analytics.hpp"

#include <vector>

namespace algoforge::analytics {

std::vector<WalkForwardFold> walkforward_anchored(
    int        n_bars,
    int        n_folds,
    int        train_min_bars,
    BacktestFn backtest_fn
) {
    std::vector<WalkForwardFold> results;
    if (n_folds <= 0 || n_bars <= train_min_bars) return results;

    int remaining     = n_bars - train_min_bars;
    int fold_size     = remaining / n_folds;   /* integer division */
    if (fold_size <= 0) return results;

    for (int k = 0; k < n_folds; ++k) {
        int train_end  = train_min_bars + k * fold_size;
        int test_start = train_end;
        int test_end   = test_start + fold_size;
        if (test_end > n_bars) test_end = n_bars;
        if (test_start >= test_end) break;

        WalkForwardFold fold;
        fold.fold_index  = k;
        fold.train_start = 0;
        fold.train_end   = train_end;
        fold.test_start  = test_start;
        fold.test_end    = test_end;

        auto [tm, em] = backtest_fn(0, train_end, test_start, test_end);
        fold.train_metrics = tm;
        fold.test_metrics  = em;

        results.push_back(fold);
    }

    return results;
}

std::vector<WalkForwardFold> walkforward_rolling(
    int        n_bars,
    int        train_window,
    int        test_window,
    int        step,
    BacktestFn backtest_fn
) {
    std::vector<WalkForwardFold> results;
    if (train_window <= 0 || test_window <= 0 || step <= 0) return results;

    int fold_index = 0;
    for (int i = 0; i + train_window + test_window <= n_bars; i += step) {
        int train_start = i;
        int train_end   = i + train_window;
        int test_start  = train_end;
        int test_end    = test_start + test_window;
        if (test_end > n_bars) test_end = n_bars;
        if (test_start >= test_end) break;

        WalkForwardFold fold;
        fold.fold_index  = fold_index++;
        fold.train_start = train_start;
        fold.train_end   = train_end;
        fold.test_start  = test_start;
        fold.test_end    = test_end;

        auto [tm, em] = backtest_fn(train_start, train_end, test_start, test_end);
        fold.train_metrics = tm;
        fold.test_metrics  = em;

        results.push_back(fold);
    }

    return results;
}

} /* namespace algoforge::analytics */
